processing/src/main/java/org/apache/druid/query/filter/TypedInFilter.java

```diff
@@ -0,0 +1,713 @@
+    return column.equals(that.column) &&
+           Objects.equals(matchValueType, that.matchValueType) &&
+           compareValues(lazyMatchValues.get(), that.lazyMatchValues.get(), matchValueType) &&
+           Objects.equals(filterTuning, that.filterTuning);
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(lazyMatchValues.get(), column, matchValueType, filterTuning);
+  }
+
+  private byte[] computeCacheKey()
+  {
+    // Hash all values, in sorted order, as their length followed by their content.
+    final Hasher hasher = Hashing.sha256().newHasher();
+    for (Object v : lazyMatchValues.get()) {
+      if (v == null) {
+        // Encode null as length -1, no content.
+        hasher.putInt(-1);
+      } else {
+        final String s = Evals.asString(v);
+        hasher.putInt(s.length());
+        hasher.putString(s, StandardCharsets.UTF_8);
+      }
+    }
+
+    return new CacheKeyBuilder(DimFilterUtils.NEW_IN_CACHE_ID)
+        .appendString(column)
+        .appendByte(DimFilterUtils.STRING_SEPARATOR)
+        .appendString(matchValueType.asTypeString())
+        .appendByte(DimFilterUtils.STRING_SEPARATOR)
+        .appendByteArray(hasher.hash().asBytes())
+        .build();
+  }
+
+  private static boolean checkSorted(List<?> unsortedValues, ColumnType matchValueType)
+  {
+    final Comparator<Object> comparator = matchValueType.getNullableStrategy();
+    Object prev = null;
+    boolean needsCoerceCheck = true;
+    for (Object o : unsortedValues) {
+      if (needsCoerceCheck && o != null) {
+        Object coerced = coerceValue(o, matchValueType);
+        //noinspection ObjectEquality
+        if (coerced != o) {
+          return false;
+        }
+        needsCoerceCheck = false;
+      }
+      if (prev != null && comparator.compare(prev, o) >= 0) {
+        return false;
+      }
+      prev = o;
+    }
+    return true;
+  }
+
+  @Nullable
+  private static Object coerceValue(@Nullable Object o, ColumnType matchValueType)
+  {
+    if (o == null) {
+      return o;
+    }
+    switch (matchValueType.getType()) {
+      case STRING:
+        return DimensionHandlerUtils.convertObjectToString(o);
+      case LONG:
+        return DimensionHandlerUtils.convertObjectToLong(o);
+      case FLOAT:
+        return DimensionHandlerUtils.convertObjectToFloat(o);
+      case DOUBLE:
+        return DimensionHandlerUtils.convertObjectToDouble(o);
+      default:
+        throw InvalidInput.exception("Unsupported matchValueType[%s]", matchValueType);
+    }
+  }
+
+  private static List<?> sortValues(List<?> unsortedValues, ColumnType matchValueType)
+  {
+    final Object[] array = unsortedValues.toArray(new Object[0]);
+    // check if values need coerced
+    for (int i = 0; i < array.length; i++) {
+      Object coerced = coerceValue(array[i], matchValueType);
+      //noinspection ObjectEquality
+      if (coerced != null && array[i] == coerced) {
+        // assume list is all same type objects...
```

Comment:
Is this going to cause issues when reading JSON that has `values` set rather than `sortedValues`? (Where they might not all be the same type)

Also, do we care? I suppose if we aren't documenting this, then there's no way we'd get JSON with `values` in it. The serializer always writes `sortedValues`. You know, though, if this is the thinking, then we should really also split the constructor into two: a `@JsonCreator` that takes `sortedValues` only (not `values`) and a non-Jackson-enabled constructor that accepts possibly-unsorted values. No sense in supporting `values` in the JSON if we aren't going to use it.

Reply:
hmm, yea I guess i could always coerce while sorting just to be safe, was trying to short-circuit if it looked like it wasnt going to be necessary and was just assuming people wouldn't troll the set with mixed type junk like strings and numbers over json (and a mix of like ints and longs wouldn't really be a problem because the column type comparators handle the stuff as numbers rather than exact types to guard against json serde shenanigans)

i guess i was still on the fence of whether or not we document. we did document the other new native typed filters, so was sort of assuming we would this one too, especially since there is a performance improvement

Reply:
added coercion while sorting just in case